### DISCUSSION TITLE
URLs from storage that don’t expire

### DISCUSSION THREAD

**Discussion by zaynelovecraft** [supabase/supabase#7626] [Questions]
I have a CMS dashboard for admin to upload pictures to storage bucket. Can I generate signedURLs with a 10 year expiry time or forever (does it have to be set in seconds?). Then can I store that url in a table so I can query the table for all image urls. I am using next js. For instance a year is 31,536,000 seconds. so is this correct? - .createSignedUrl('folder/avatar1.png', 31,536,000) I need the urls to last until deleted. I get urls with Next js Get static props ISR. Or is there a better way to do this ? Please help Thank you!

**Comment by GaryAustin1** [comment]
Yes you can set them for a very long time.  
The token that gets attached is a jwt, for example:

![image](https://user-images.githubusercontent.com/54564956/178069994-e1356499-99e7-4470-bbc2-2159299d8222.png)
Which shows an expiring time 10 years from now.

You could also get the same thing (relevant only for very long signedUrl) using a public url with a hard to guess url by attaching -[uuid] or -[time in msec] to the filename.

Both if discovered can be used by anyone and a uuid or the exact time plus a filename are pretty close to impossible to guess.  The advantage with a public url is you can pre-generate it and don't have to make an extra getSignedUrl request to the SB server (public URL is calculated so does not need a server request).

**Comment by zaynelovecraft** [comment]
How would I pre generate public urls. Say I upload the photo to a storage bucket and at the same time want to then store the url in a table. What’s the simplest way to do this ?

**Comment by GaryAustin1** [comment]
You can use .GetPublicUrl even if you have not stored the file. It just builds a path based on your instance URL, some constants, the bucket name and pathname. It does not make a server call.

 If you look at the URL it is pretty obvious you could also just generate with string concatenation.  

A public bucket has all the normal RLS rules of a private one, except for allowing the public URL access.

**Reply by zaynelovecraft** [reply]
Can you explain the best way to upload a photo to storage, then store the url in the a table then on my app for the hero image I use get static props then pass the url to the hero image and put the url as the SRC. What I’m trying to accomplish is basically a way for admin of website can edit content on the home page and other pages by clicking a edit button if admin is logged in then choosing a image from storage then by doing so it should set in the table the image url they chose. Srry if this isn’t good enough explanation. I would really appreciate it if I could share my screen with you. Do you have discord ?

**Reply by GaryAustin1** [reply]
I don't see any issues with the general idea you have.  You store the image to storage to make sure it uploads successfully using storage.from(bucket).upload().  Then generate the public URL and store it where you want.  Make sure the bucket is checked "public".  
Then you can move that URL around to different tables, put it in img src="url" and all should be fine.
